### PR TITLE
[SYSTEMML-1491] Add different ReLU variants ELU

### PR DESCRIPTION
**Pull request by krishnakalyan3**
- Added `ELU` activation
- Added Reference Paper
- Added `forward` test
- Added `backward` test

@niketanpansare @dusenberrymw could you please review this?.

### CODE REVIEW

**Comment by akchinSTC** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/<h2>Failed Tests: <span class='status-failure'>7</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>7</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.append/AppendMatrixTest/testAppendInBlock2SparseMR/'><strong>org.apache.sysml.test.integration.functions.append.AppendMatrixTest.testAppendInBlock2SparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.append/RBindCBindMatrixTest/testCBindDenseMR/'><strong>org.apache.sysml.test.integration.functions.append.RBindCBindMatrixTest.testCBindDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.append/RBindCBindMatrixTest/testRBindDenseMR/'><strong>org.apache.sysml.test.integration.functions.append.RBindCBindMatrixTest.testRBindDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.reorg/VectorReshapeTest/testVectorReshapeDenseMR/'><strong>org.apache.sysml.test.integration.functions.reorg.VectorReshapeTest.testVectorReshapeDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.unary.matrix/RemoveEmptyTest/testRemoveEmptyRowsDenseMR/'><strong>org.apache.sysml.test.integration.functions.unary.matrix.RemoveEmptyTest.testRemoveEmptyRowsDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.unary.matrix/RemoveEmptyTest/testRemoveEmptyRowsDenseMRwIdx/'><strong>org.apache.sysml.test.integration.functions.unary.matrix.RemoveEmptyTest.testRemoveEmptyRowsDenseMRwIdx</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2252/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.unary.matrix/RemoveEmptyTest/testRemoveEmptyRowsMultipleDenseMR/'><strong>org.apache.sysml.test.integration.functions.unary.matrix.RemoveEmptyTest.testRemoveEmptyRowsMultipleDenseMR</strong></a></li></ul>

**Comment by krishnakalyan3** [comment]
Test this Jenkins

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2255/

**Comment by niketanpansare** [comment]
Thanks Krishna. Can you also add gradcheck tests ?

File: scripts/nn/layers/elu.dml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+/*
+ * Exponential Linear Units (ELUs) nonlinearity layer.
+ */
+
+forward = function(matrix[double] X, int alpha)
+    return (matrix[double] out) {
+  /*
+   * Computes the forward pass for a ELUs nonlinearity layer.
+   *
+   * Performs an element-wise evaluation of `f(input) = max(0, input) + min(0, alpha ∗ (exp(input) − 1))`.
```
Comment (dusenberrymw) [inline scripts/nn/layers/elu.dml]: Can you check that the line length here is <= 72?  For the `nn` library, I've kept all docstring lines <= 72 characters in length.  Also, can you (1) reference the ELU paper (https://arxiv.org/abs/1511.07289v1), and (2) add documentation that this formulation matches the if/else formula from the paper `(f(x) = x if x ≥ 0 else α (exp(x) − 1))`.

**Comment by krishnakalyan3** [comment]
@niketanpansare, @dusenberrymw thank for your feedback. Will make the necessary changes shortly.

**Comment by krishnakalyan3** [comment]
@niketanpansare @dusenberrymw have implemented the gradient checks for `ELU` now and will tackle `PReLU` in a separate PR.

The functions and the gradients were compared against `pytorch` [here](https://github.com/krishnakalyan3/SystemML-Utils/blob/master/notebooks/ELU%20GradCheck.ipynb).

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2272/

**Comment by niketanpansare** [comment]
LGTM wrt my comments :+1:  ... @dusenberrymw, unless you object, I will merge this next week.

**Comment by dusenberrymw** [comment]
@krishnakalyan3 Thanks for adding tests.  The forward function test compared against PyTorch looks great.  However, can you please add a random automated gradient test in the `nn/test/gradcheck.dml` in the style of those tests?  The overall idea is that the forward functions can be compared to another implementation for correctness, and then the gradient tests can be automated for random values to ensure that the backward functions are correct w.r.t. their corresponding forward functions.

> Add gradient tests in `nn/test/gradcheck.dml`. The ReLU test can be copied and adapted for the ELU function, and the affine layer test could be copied and adapted for the PReLU. These will ensure that the backward functions are computing the correct derivatives.

**Comment by krishnakalyan3** [comment]
@dusenberrymw added gradient checks.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2274/

**Comment by dusenberrymw** [comment]
@krishnakalyan3 Sorry for the delay on this.  One last thing and this can be merged (pending passing tests): please enable the tests in the [`nn/test/run_tests.dml`](https://github.com/apache/systemml/blob/master/scripts/nn/test/run_tests.dml) script by adding one line to call the grad check test in the grad check section, and one line to call the other test in the other test section.  Both sections are in alphabetical order.

Given that, you should be able to run `../bin/systemml nn/test/run_tests.dml` from the `scripts` folder and look for any failures via `ERROR...` messages.  (Side note: the print statements from these tests used to be in order, but a somewhat recent change caused that to no longer be the case -> cc @mboehm7 as a reminder).  The tests will also be run automatically from Java via the [`NNTest.java`](https://github.com/apache/systemml/blob/master/src/test/java/org/apache/sysml/test/integration/scripts/nn/NNTest.java) test case once the above 2 lines are added to `run_tests.dml`.

**Comment by mboehm7** [comment]
@dusenberrymw well I thought I fixed it immediately after the 1.0 release by always sorting prints within a DAG according to their line numbers. Anyway, now I know what you mean: after function inlining and merge of DAGs, we end up sorting the prints according to their own line numbers not the line numbers of the original function calls. I'll take care of it.

**Comment by krishnakalyan3** [comment]
gentle ping @dusenberrymw. Thank you for the reviews so far :).

File: scripts/nn/layers/elu.dml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+   *  - dout: Gradient wrt `out` from upstream, of same shape as `X`.
+   *  - X: Previous input data matrix, of shape (any, any).
+   *  - alpha: Minimum value that the ELU can reach
+   *           Typical value 1
+   *
+   * Outputs:
+   *  - dX: Gradient wrt `X`, of same shape as `X`.
+   */
+   out = max(0, X) + min(0, alpha * (exp(X) - 1))
+   dX = ((X > 0) + min(0, alpha + out)) * dout
```
Comment (dusenberrymw) [inline scripts/nn/layers/elu.dml]: I believe this needs to be updated to `dX = ((X > 0) + (X < 0)*(alpha*exp(X))) * dout`.

Comment (krishnakalyan3) [reply scripts/nn/layers/elu.dml]: @dusenberrymw thank you for catching this.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2306/

**Pull request closed without merging**